src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+  let updatedState = { ...state };
+
+  for (const action of actions) {
+    updatedState = { ...updatedState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(updatedState, action.extraData);
+        statesArray.push(updatedState);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete updatedState[key];
+        }
+        statesArray.push(updatedState);
+        break;
+
+      case 'clear':
+        for (const key in updatedState) {
+          delete updatedState[key];
+        }
+        statesArray.push(updatedState);
```

Comment:
Again, clone `updatedState` before pushing it to `statesArray` to ensure each state is independent of others.